### PR TITLE
Add Micrometer Support for Consumer and Producer

### PR DESCRIPTION
**Pull request by garyrussell**
Resolves https://github.com/reactor/reactor-kafka/issues/206#issuecomment-1006119883

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -127,8 +128,10 @@ configure(allprojects) { project ->
 
   dependencies {
     compileOnly "com.google.code.findbugs:jsr305:$googleJsr305Version"
+    compileOnly "io.micrometer:micrometer-core:$micrometerVersion"
```
Comment (artembilan) [inline build.gradle:131]: Not `optional`?

File: src/main/java/reactor/kafka/receiver/ImmutableReceiverOptions.java [hunk 2]
```diff
@@ -30,13 +38,6 @@
 import java.util.regex.Pattern;
 import java.util.stream.Collectors;
 
-import org.apache.kafka.clients.consumer.ConsumerConfig;
-import org.apache.kafka.common.TopicPartition;
-import org.apache.kafka.common.config.ConfigException;
-import org.apache.kafka.common.serialization.Deserializer;
-import reactor.core.scheduler.Scheduler;
-import reactor.core.scheduler.Schedulers;
```
Comment (artembilan) [inline src/main/java/reactor/kafka/receiver/ImmutableReceiverOptions.java:38]: Is it correct for this project to rearrange imports?

Not related, but in Spring Integration 6.0 I'm going to follow IDE settings from Spring Framework: https://github.com/spring-projects/spring-integration/issues/3686
From where I believe the import order is exactly like it was before your refactoring here.

**Review by artembilan — CHANGES_REQUESTED** [review]

File: build.gradle (hunk 1 above)
Comment (garyrussell) [reply build.gradle:131]: This version of gradle doesn't seem to support it; we really need to upgrade to a recent version but my gradle/groovy skills are not up to it.

File: src/main/java/reactor/kafka/receiver/ImmutableReceiverOptions.java (hunk 2 above)
Comment (garyrussell) [reply src/main/java/reactor/kafka/receiver/ImmutableReceiverOptions.java:38]: We had nothing like that when this syntax used to work 

    compile ("com.fasterxml.jackson.core:jackson-core:$jacksonVersion", optional)

https://github.com/spring-projects/spring-kafka/blob/5c54ad02339ab1849f98fd3e785f811c6deb17c5/build.gradle#L198

Comment (artembilan) [reply src/main/java/reactor/kafka/receiver/ImmutableReceiverOptions.java:38]: Yes. I don't mind, but turns out Gradle has decided to go further and above limited number of Maven dependency variants.
But unfortunately they liquidated an `optional` as out-of-the-box at all...

**Review by artembilan — APPROVED** [review]
LGTM.

Still don't have permissions to merge PRs in this project 😄

File: build.gradle (hunk 1 above)
Comment (garyrussell) [reply build.gradle:131]: Yes but a lot has changed since then `compile` etc have been deprecated - there are other changes too - e.g. maven publishing.

Comment (simonbasle) [reply build.gradle:131]: @garyrussell @artembilan upgrading this project to Gradle 7.4 is on my todo list. made a bit difficult by the fact that it lagged on 5.x indeed...

File: src/main/java/reactor/kafka/receiver/ImmutableReceiverOptions.java (hunk 2 above)
Comment (simonbasle) [reply src/main/java/reactor/kafka/receiver/ImmutableReceiverOptions.java:38]: in core we do use `compileOnly` (but indeed these don't appear in the pom, see https://github.com/reactor/reactor-core/issues/2262)

**Comment by garyrussell** [comment]
Need to rework this on Monday due to gradle upgrade (thanks @simonbasle ).

**Review by artembilan — APPROVED** [review]

**Review by simonbasle — COMMENTED** [review]
nit: added `maxDeferredCommits` to equals and hashCode looks like a separate issue.

I was also going to bring up that with upcoming Micrometer 2.0, maybe there's little sense in adding Micrometer 1 metrics so late in the cycle. But that's not going to be GA until the end of the year, and given the metrics code in this change I don't think the migration to Micrometer 2 will be terribly hard. so 👍

**Comment by garyrussell** [comment]
Another thought would be to just add the listener interface and leave off the micrometer implementations out and just add the code to the reference manual since it's so small; the user can copy/paste.

WDYT?

**Comment by migroskub** [comment]
> Another thought would be to just add the listener interface and leave off the micrometer implementations out and just add the code to the reference manual since it's so small; the user can copy/paste.
> 
> WDYT?

I think it misses the whole point. As a user, I'd like to just have an easy way of making this work. If you expect most of the users to copy-paste the metrics, I think it worth it to let them be there, and the more advanced needs will implement the api.

**Comment by simonbasle** [comment]
@garyrussell yeah that could be an intermediate way to improve things in 1.3.x.

Let me list a few factoids and constraints:
 - Micrometer 2.x is around the corner, GA late 2022
 - there are no metrics at all in reactor-kafka 1.3.x right now
 - reactor-kafka 1.3.x cannot use Micrometer 2.x (we need to maintain compatibility with reactor-core 3.4.x which uses  Micrometer 1)
 - there could be a reactor-kafka 1.4.0, part of the 2022.0 upcoming release cycle (or compatible with it) that uses Micrometer 2.x (late 2022)

@migroskub I understand your perspective, but as stated above there are _no_ metrics right now. if you're adamant about metrics _right now_ then you could be considered an advanced user and would get an interface to listen and implement metrics with some effort, albeit minimal.
Then in 1.4.0 all users will get something more future-proof working off the shelf.

Just playing devil's advocate here. I don't want to rush implementing metrics with Micrometer 1.x when Micrometer 2.x is around the corner.

**Pull request closed without merging**